[WC-897] Rewrite rich text to pluggable

raffiar opened the pull request:
## Checklist
- Contains unit tests ❌
- Contains breaking changes ✅
- Contains Atlas changes ❌
- Compatible with: MX 9️⃣

#### Web specific
- Contains e2e tests ❌
- Is accessible ✅
- Compatible with Studio ✅
- Cross-browser compatible ✅

#### Feature specific
- Comply with designs ✅
- Comply with PM's requirements ✅

## This PR contains
- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Other (describe)

## What is the purpose of this PR?
Rewrite the rich text editor to pluggable widget. We changed the core library from `Quill` to `CKEDITOR`, implemented new configurations and toolbar settings.
This is in general working version for customers, main functionality will not be changed in next commit messages.
## Relevant changes
Complete rewrite of rich text, new library is presented as a core. Moved from custom widgets to pluggable. `README` and `CHANGELOG` still in progress, will be presented in a next commit messages.
## What should be covered while testing?
If it works as expected on client side, if it has every item except those which marked **red**, if multiple editors in same page works correct.if toolbar items works as expected.
## Extra comments (optional)
UPDATE: Rollup configuration adde

raffiar committed:
feat(rich-text-web): move files and folders to pluggable

raffiar committed:
feat(rich-text-web): correct structure, fixed dependencies

raffiar committed:
feat(rich-text-web): added consistency checks

raffiar committed:
feat(rich-text-web): added contentFiltering properties

raffiar committed:
feat(rich-text-web): basic configurations and property handling

raffiar committed:
feat(rich-text-web): move editor instance to another component

raffiar committed:
feat(rich-text-web): missing properties

raffiar committed:
feat(rich-text-web): define local presets

raffiar committed:
feat(rich-text-web): define all presets and advanced configurations

raffiar committed:
feat(rich-text-web): simplfy flow, remove unnecessary props

raffiar committed:
feat(rich-text-web): group advanced toolbar items by their group name

raffiar committed:
feat(rich-text-web): working version with cdn

raffiar committed:
feat(rich-text-web): rolllup bundling configuration

raffiar committed:
feat(rich-text-web): added important image files from library to src

- removed  not defined properties
- removed class property
- commented rollup configuration which was replacing icon path

Keraito commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorPresets.ts:
```diff
@@ -0,0 +1,156 @@
+            config.removeButtons = "";
+            /* temporary removed exportpdf*/
+            config.extraPlugins =
+                "save,templates,newpage,print,forms,find,selectall,copyformatting,div,divarea,justify,bidi,language,font,colorbutton,showblocks";
```
Is this suddenly lowercase compared to the other switch cases intentional?

raffiar replied:
I didn't understand the question, but the case is defined by CKEditor

Keraito replied:
In L94 and L101 everything is camelcased: "Save", "Source", "SelectAll", but here it's just lowercased "save" "source"  "selectall". That seems a bit strange.

raffiar replied:
Again, it's not controlled by me, it's ckeditor.
 CamelCased items are definition for toolbar group (Buttons). lowercase items are plugin names. They are matching but it's case sensitive.
 https://github.com/ckeditor/ckeditor4/blob/master/plugins/about/plugin.js check the link, it's an example of one of the plugins in ckeditor

Keraito replied:
Okay, it wasn't clear that this is about plugins and the other cases are about removeButtons as the code is very similar 😓  if you don't think there's a more structured way to implement this, let's keep it like this i guess

Keraito commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorPresets.ts:
```diff
@@ -0,0 +1,156 @@
+            toolbarGroup.splice(6, 0, "/");
+            config.toolbarGroups = toolbarGroup;
+            config.removeButtons = "";
+            /* temporary removed exportpdf*/
```
It's not really clear to me what this mean:
- Why do we need to do this?
- Temporary until when?

raffiar replied:
`exportpdf` toolbar item was presented on PM's document, but due to some additional library connections I hide it from the list If we decide to include that plugin I will bring it back, if not, i'll just remove the comment.

`exportpdf` requires some authentication token, which should be defined by user itself, so it can be advanced configuration @cjfhodges

Keraito replied:
so this is temporary that it still needs to be fixed from our side? If so, please prefix the comment with `TODO:`

Keraito commented on packages/pluggableWidgets/rich-text-web/package.json:
```diff
@@ -14,39 +14,42 @@
-    "quill": "^1.3.6",
-    "sanitize-html": "^2.0.0"
+    "sanitize-html": "^2.6.0",
+    "ckeditor4-react": "^2.1.0",
+    "ckeditor4": "^4.17.1"
```
I know there was a reason, but I don't remember it clearly: 
- Why do we need both `ckeditor` and `ckeditor-react`?
- What is `ckeditor` used for and what is `ckeditor-react` used for?

iobuhov replied:
- `ckeditor` - the editor code itself (we going to copy bundle from this package + plugins)
- `ckeditor-react` - the react wrapper that actually manage `CKEditor` instance lifecycle

diego-antonelli replied:
Isn't ckeditor4 a dependency of ckeditor4-react?

iobuhov replied:
No
```sh
❯ npm info ckeditor4-react dependencies
{ 'ckeditor4-integrations-common': '^1.0.0', 'prop-types': '^15.7.2' }
```

Keraito replied:
🤯

diego-antonelli replied:
Then I am resolving this discussion

Keraito commented on packages/pluggableWidgets/rich-text-web/src/components/RichText.tsx:
```diff
@@ -0,0 +1,111 @@
+            style={{ width, height }}
+        >
+            <div ref={setElement} id={props.name} />
+            <MainEditor config={ckeditorConfig} key={key} />
```
I don't quite get how the actual editor is actually rendered onto the screen. We store the above div element in state and set some CKEditor specific configurations. In `MainEditor`, we use the CKEditor hook, but it doesn't return or render anything as far as I see. So what is responsible for putting the CKEditor on the screen?

raffiar replied:
In `MainEditor` component it destroys `CKEditor` instance and creates a new one (``` useCKEditor(config);```)

here, useCKEditor generates ckeditor instance every time when some of the props are changed.

It's looking for div `(<div ref={setElement} id={props.name} />)` and placing it with an `element` property which is passed through `ckeditorConfig` object

[element  - Reference to underlying to DOM element.](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_react_v2.html#api-2)

Why we need to destroy and initialise ckeditor again and again? Because it doesn't provide api to update config object dynamically.

`stringAttribute` is a dynamic property and it has status (pending, completed) or something like that, (maybe you can explain why it's working in this way), but here the problem was, when you set `readOnly` config `false` on your editor props, u always receive true at the beginning and then false at the end. And because there is no option to dynamically change readOnly value on CKEditor you need to destroy and initialise again.

I hope I answered sufficiently

Keraito replied:
So https://github.com/mendix/widgets-resources/pull/1125/files#diff-85d3cc8ae664c556ca3ca667615281345e689ec33861bdd6ff56cc051b1ae809R92 is responsible for rendering the CKEditor onto the `element`?

raffiar replied:
Yes, if you look at the initial definition you will see that I am passing `element` property there. Element is a ref element from div

Keraito commented on packages/pluggableWidgets/rich-text-web/src/components/__tests__/RichText.spec.tsx:
```diff
@@ -0,0 +1,38 @@
+    });
+    it("render DOM structure", () => {
+        const { asFragment } = render(<RichTextEditor {...defaultRichTextProps} />);
+        expect(asFragment()).toMatchSnapshot();
```
Could you add tests that verify inserting certain widget properties/component props result into the proper configurations in CKEditor?

Keraito reviewed (changes requested):
- Have a few questions about different parts as I don't understand them fully. 
- Also I think we should get rid of the `any` typings as much as possible, as they're not really adding anything.
- Some remarks about Studio stuff in editorConfig.

Keraito commented on packages/pluggableWidgets/rich-text-web/CHANGELOG.md:
```diff
@@ -5,6 +5,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+- We added a lot of cool features such as code highlight, allow-disallow content filtering, customizable toolbar, general toolbar presets (**Basic**, **Advanced**, **Full**).
+- We added read only style modes.
+- We added editor style mode **(Toolbar | Inline)**.
+### Changed
```
linebreak before title

Keraito commented on packages/pluggableWidgets/rich-text-web/CHANGELOG.md:
```diff
@@ -5,6 +5,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+- We added read only style modes.
+- We added editor style mode **(Toolbar | Inline)**.
+### Changed
+- We changed the core library for Rich Text. Moved from Quill to CKEditor4.
```
Is it worth to get so specific about technical details @diego-antonelli?

Keraito commented on packages/pluggableWidgets/rich-text-web/rollup.config.js:
```diff
@@ -0,0 +1,75 @@
+
+        /* Need to replace icon path from bundled js files, it's step 0 and 1, 
+        tried to find much safer comparison but there are no uniqueness, so the best idea is to replace the path by index */
+        /* if (index === 0 || index === 1) {
+            config.plugins.push(replaceIconPath());
+        }*/
```
Commented code

raffiar committed:
feat(rich-text-web): added unit tests

raffiar commented on packages/pluggableWidgets/rich-text-web/rollup.config.js:
```diff
@@ -0,0 +1,72 @@
+        }
+        /* this step is required by sanitize-html library */
+        config.plugins.push(json());
+        return { ...config };
```
The only reason is to keep object immutable

iobuhov replied:
I think there should be no semantic deference between `config` and `{ ...config }`, so I would stick with last one

diego-antonelli replied:
@raffiar can you then change to `config`?

raffiar replied:
If it will not affect other configs then yes, because now I see `null` items in `config.plugins` array, which makes me think that object mutation is not a good idea here.

raffiar commented on packages/pluggableWidgets/rich-text-web/rollup.config.js:
```diff
@@ -0,0 +1,72 @@
+                    /* we need to empty every single css inside the folder to avoid duplicate bundling,
+                     because if the file even not imported anywhere it will be compiled inside RichText.js and RichText.mjs */
+                    if (extname(src) === ".css") {
+                        return through((chunk, enc, done) => {
+                            done(null, "");
+                        });
```
We faced file loading issue in the process of integration. **CDN vs local path.**
When we defined `CKEDITOR` path to look in a local source it started to load files automatically from local, and meantime same css files was being loaded by rollup into bundled **main.js** and **main.mjs** files.

The idea behind this configuration is to rewrite every `.css` file during compilation process and replace the content with emptiness .

With this way we will avoid duplications in our code, also network will not give error about not founded files. because the other option was not to include that files in dist folder at all.

raffiar replied:
Correct, you can check [recursive-copy](https://www.npmjs.com/package/recursive-copy) library examples for more examples

raffiar commented on packages/pluggableWidgets/rich-text-web/rollup.config.js:
```diff
@@ -0,0 +1,72 @@
+                    "**/ckeditor.js",
+                    "**/config.js",
+                    "**/styles.js",
+                    "**/README.md",
+                    "**/SECURITY.md",
```
I think these 2 files are not important, but we should ask @diego-antonelli and @cjfhodges if there are any files that need to be included in the bundle because of `LICENSE` type

diego-antonelli replied:
I think not even LICENSE is required as the new rollup-plugin will collect the dependencies and add them into dependencies.txt inside the widget mpk.

raffiar commented on packages/pluggableWidgets/rich-text-web/src/RichText.editorPreview.tsx:
```diff
@@ -0,0 +1,32 @@
+    className: string;
+    readOnly: boolean;
+    name: string;
+}
```
Done

raffiar commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorConfigs.ts:
```diff
@@ -0,0 +1,82 @@
+import { SET_PRESET, TOOLBAR_GROUP, ToolbarGroup } from "./ckeditorPresets";
+
+interface Plugins {
+    codesnippet: any;
```
this one was mistake, it should be `type` not `interface`. So i'll change it

raffiar commented on packages/pluggableWidgets/rich-text-web/src/ui/RichText.scss:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+
+    &.editor-borderedToolbar {
```
For now, we don't. But it's an empty representation of what possibly can be applied to container div. I'll prefer to keep it

diego-antonelli replied:
Do we need it for something?

Keraito replied:
If not, just remove it.

raffiar commented on packages/pluggableWidgets/rich-text-web/src/ui/RichText.scss:
```diff
@@ -0,0 +1,40 @@
+@use "../../../node_modules/ckeditor4/plugins/tableselection/styles/tableselection.css";
+@use "../../../node_modules/ckeditor4/plugins/dialog/styles/dialog.css";
+@use "../../../node_modules/ckeditor4/plugins/copyformatting/styles/copyformatting.css";
+@use "./Scayt";
```
Basically it's because of namespace. there are multiple `dialog.css` files in the import. But, what I did I just added different namespaces. Didn't know that scss has `as`

raffiar commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorPresets.ts:
```diff
@@ -0,0 +1,156 @@
+            break;
+        case "standard":
+            toolbarGroup = [...TOOLBAR_GROUP];
+            toolbarGroup.splice(4, 0, "/");
```
it's basically adding new line into toolbar items.
Check the link and move "/" from one index to another
https://ckeditor.com/latest/samples/toolbarconfigurator/index.html#advanced

Keraito replied:
https://github.com/mendix/widgets-resources/pull/1125#discussion_r773066820 Same thing

raffiar commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorPresets.ts:
```diff
@@ -0,0 +1,156 @@
+    let toolbarGroup: Array<ToolbarGroup | string> = [];
+    switch (type) {
+        case "basic":
+            config.toolbarGroups = [...TOOLBAR_GROUP];
```
same as in `...config` . I want to keep original object untouched

Keraito replied:
as mentioned before, you can just use `config`

Keraito replied:
Same comment about `config` as others

raffiar replied:
You cant, component renders every time when some props changed.
it means every time it will try to fetch correct preset with already modified objects.

Not for the case of "basic", because we don't modify anything here, but for the other ones, because we push "/" in specific indexes

Keraito replied:
Can you explain why "you can't"? Based on your comment that it re-renders everytime a prop changes, your current code would mean it always re-renders because everytime this code is ran, toolbarGroups will become a new reference because you're spreading into a new array, which would be a "change".

raffiar replied:
I don't follow you. the second part of your answer is confusing me, but i'll give a short code example to make it clear why I am spreading.

Example 1. Result without spreading. Main object modifies every time when you do some actions
```
const initialArray = [];

const result = () => {
  const modifiedArray = initialArray;
  modifiedArray.push("value");
  return modifiedArray;
}

const result1 = result();
const result2 = result();

console.log(initialArray, result1, result2);
['value', 'value'], ['value', 'value'], ['value', 'value']
```

So as you can see in our case it will push `"/"` every time when react component renders.

Example 2. When spreading initialObject it creates new object without reference so you can push your changes to the new array and always have your desired result.

```
const initialArray = [];

const result = () => {
  const modifiedArray = [...initialArray];
  modifiedArray.push("value");
  return modifiedArray;
}

const result1 = result();
const result2 = result();

console.log(initialArray, result1, result2);

[] ['value'] ['value']
```

raffiar commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorPresets.ts:
```diff
@@ -0,0 +1,156 @@
+            break;
+        case "full":
+            toolbarGroup = [...TOOLBAR_GROUP];
+            toolbarGroup.splice(3, 0, "/");
+            toolbarGroup.splice(5, 0, "/");
+            toolbarGroup.splice(6, 0, "/");
```
https://github.com/mendix/widgets-resources/pull/1125#discussion_r769480338
Later, in case when we receive feedback from customer that they want to define their own  break lines, we can change index's with their inputs

Keraito replied:
Is there any meaning behind these numbers/indices? Right now, they are quite arbitrary and it will be very difficult to understand them in the future.

raffiar replied:
Have you checked the comment above?

iobuhov commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorPresets.ts:
```diff
@@ -0,0 +1,146 @@
+    { name: "about", items: ["About"] }
+];
+
+function SET_PRESET(type: "basic" | "standard" | "full"): CKEditorConfig {
```
NIT: I think it's bit confusing to see function name as upper case. I agree that this is "constant" value, but I think it's better to rename it to `setPreset`

iobuhov commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorPresets.ts:
```diff
@@ -0,0 +1,146 @@
+    return config;
+}
+
+function SET_CUSTOM(groups: string[], withGroupNames = true): CKEditorConfig {
```
Rename it to `createCustomPreset`?

raffiar commented on packages/pluggableWidgets/rich-text-web/src/RichText.tsx:
```diff
@@ -0,0 +1,75 @@
+        if (preset !== "custom") {
+            return getPreset(preset);
+        } else {
+            const groupKeys = Object.keys(props).filter((key: string) => (key.includes("Group") ? key : null));
```
I will prefer to keep it dynamic, making static list of item is a bad idea, cause in different cases the number of items can be much more for example.

`groupKeys` now just making sure that you are passing correct group names to 'createCustomToolbar' method. But in general, if you add something which is not in the `ToolbarItems` it will be excluded from final list.

Keraito replied:
> cause in different cases the number of items can be much more for example.

What do you mean exactly and how does that make a static list bad?

> it will be excluded from final list.

Can you point me to the code that does that? Because right now I only see this filtering, which only checks for "Group".

raffiar replied:
My point is it's better to get any props dynamically if you can, rather than hardcoding the staff.

And here is the part when I am checking if the item is in the list
[packages/pluggableWidgets/rich-text-web/src/utils/ckeditorPresets.ts](https://github.com/mendix/widgets-resources/pull/1125/files#diff-1f1184f8a3abe3737a7db00c3403ec989aa59ef2dcb07ba2951a516e391741b9R124)

In any case, if in the future, someone by mistake add another parameter which includes "Group" in the name, it will be excluded from final list.

raffiar committed:
feat(rich-text-web): comment fixes, improvements

raffiar committed:
feat(rich-text-web): fix failing readOnly property

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/RichText.editorConfig.ts:
```diff
@@ -0,0 +1,65 @@
+    }
+    if (platform === "web") {
+        transformGroupsIntoTabs(defaultProperties);
+        hidePropertyIn(defaultProperties, values, "advancedMode");
```
I think this check is incorrect. We should only show advanced mode in platform web and not the opposite.

diego-antonelli replied:
I think you misunderstood. You should hide this property when is not web, so in this case this line should be moved to the else. Because we dont want the toggle for advanced mode in Studio Pro.

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/RichText.editorConfig.ts:
```diff
@@ -0,0 +1,65 @@
+    return defaultProperties;
+}
+
+export function check(_values: RichTextPreviewProps): Problem[] {
```
If this functions is not being used, then better remove.

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/RichText.xml:
```diff
@@ -0,0 +1,303 @@
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
+    <name>Rich Text</name>
+    <description>Rich inline or toolbar text editing</description>
+    <icon>
```
We dont use <Icon> anymore.

diego-antonelli replied:
Missing strudioProCategory and studioCategory

raffiar replied:
Should I remove icon?

diego-antonelli replied:
Yes, it will be a file produced by Ana in the src folder.

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/RichText.xml:
```diff
@@ -0,0 +1,303 @@
+                </property>
+                <property key="sanitizeContent" type="boolean" defaultValue="true">
+                    <caption>Sanitize content</caption>
+                    <description>The strict sanitize rules will remove all unsupported HTML tags and attributes before displaying.WARNING disabling this option is a security risk for XSS attacks
```
Missing space.

raffiar replied:
where?

diego-antonelli replied:
.WARNING

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/components/MainEditor.tsx:
```diff
@@ -0,0 +1,6 @@
+import { useCKEditor, CKEditorHookProps } from "ckeditor4-react";
+
+export const MainEditor = ({ config }: { config: CKEditorHookProps<string> }): null => {
+    useCKEditor(config);
```
Cant we just use this hook instead of creating a new component without anything else relevant?

raffiar replied:
nope, https://github.com/mendix/widgets-resources/pull/1125#discussion_r772403719
We already discussed a possible solution with @iobuhov and the only way to have ckeditor working is this.

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/components/RichText.tsx:
```diff
@@ -0,0 +1,112 @@
+import sanitizeHtml from "sanitize-html";
+import classNames from "classnames";
+import { ReadOnlyStyleEnum, EnterModeEnum, ShiftEnterModeEnum, AdvancedConfigType } from "../../typings/RichTextProps";
+import { MainEditor } from "./MainEditor";
```
Use the hook instead, seems to be more readable.

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/components/RichText.tsx:
```diff
@@ -0,0 +1,112 @@
+          };
+    const [ckeditorConfig, setCkeditorConfig] = useState<CKEditorHookProps<"change">>({
+        element,
+        editorUrl: "/widgets/ckeditor/ckeditor.js",
```
Maybe we can do better and rely on `${window.mx.remoteUrl}/widgets/...` to assure the URL is correct.

raffiar replied:
how should I deal with typings error here? is there any example?

diego-antonelli replied:
You should implement the global for window with typings. Same as https://github.com/mendix/widgets-resources/blob/3dea5d093148b8e7fc1b26ed364b8a42036e013d/packages/pluggableWidgets/datagrid-date-filter-web/typings/global.d.ts#L26

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/ui/RichText.scss:
```diff
@@ -0,0 +1,40 @@
+@use "../../../node_modules/ckeditor4/contents.css";
```
`@use "~ckeditor4/contents.css";` will automatically resolve in the correct node_modules without relative path.

diego-antonelli replied:
Same applies for lines below

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorConfigs.ts:
```diff
@@ -0,0 +1,82 @@
+    const toolbarObj: {
+        [key: string]: Array<CtItemTypeEnum | "-">;
+    } = {};
+    items.forEach(item => {
```
Cant we map instead?

raffiar replied:
Maps supposed to return something, here we just modifying object outside

raffiar replied:
From functionality perspective, of course, it will work as expected, but from logical perspective it's better to `forEach` than `map` if you don't expect `return`

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorConfigs.ts:
```diff
@@ -0,0 +1,82 @@
+            items: toolbarObj[key]
+        })
+    );
+    return toolbarArray;
```
I have the feeling all lines above can be combined in one map

raffiar replied:
We can't combine it in one map, maybe nested

Input example
```
[
    {ctItemToolbar: "toolbar-1", ctItemType: "separator"},
    {ctItemToolbar: "toolbar-1", ctItemType: "About"},
    {ctItemToolbar: "toolbar-2", ctItemType: "Document"}
]
```
Desired output
```
[
    {name: "toolbar-1", items: ["About", "-"]},
    {name: "toolbar-2", items: ["Document"]}
]
```

Keraito replied:
Although I think it's possible, I don't see a quick way to get rid of the for loop from L60 to L71. But for Line 73 until 81, you can do:
```
return Object.entries(toolbarObj).map(([key, value]) => ({ name: key, items: value }));
```

raffiar replied:
So now you want me to replace 73-81 with this because of what? Just to get rid of `forEach`?

diego-antonelli commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorPresets.ts:
```diff
@@ -0,0 +1,146 @@
+                }
+            }
+        });
+        return {
```
What about return a map instead of forEaching it?

diego-antonelli reviewed (changes requested):
Good stuff! Left some improvements

Keraito commented on packages/pluggableWidgets/rich-text-web/src/RichText.editorPreview.tsx:
```diff
@@ -0,0 +1 @@
+export const preview = (): null => null;
```
Question: 
- Was it intentional to remove the entire editor preview?
- If so, what is the plan for this in the future?

raffiar replied:
It's the same case as in Charts, we need SVG

Keraito commented on packages/pluggableWidgets/rich-text-web/src/utils/ckeditorConfigs.ts:
```diff
@@ -0,0 +1,82 @@
+            items: toolbarObj[keys[j]]
+        });
+    }
+    return toolbarArray;
```
You reacted with a 👍  but didn't change anything. Did something go wrong? 🤔

raffiar replied:
Please check https://github.com/mendix/widgets-resources/pull/1125/files/d0928e73c4e47e5e1589b1cd1290f4ddb25f142b#r770602923

Keraito commented on packages/pluggableWidgets/rich-text-web/src/components/__tests__/RichText.spec.tsx:
```diff
@@ -0,0 +1,156 @@
+import { createElement } from "react";
```
You removed the DOM structure test, but we need it for backwards compatibility reasons. Can you add it back and fix it?

raffiar replied:
The is no backward compatibility, we changed the library, it renders differently.
Please correct me if I am wrong.

raffiar replied:
I don’t think it’s possible to render ckeditor structure in snapshot.
Even for the future you will not have access to ckeditor dom structure, even with that if they change version with structure your test with fail without your control.

The only snapshot which is possible here is outside div with it's classname, and I don't think that is a blocker

raffiar replied:
Added minimal snapshot testing

Keraito commented on packages/pluggableWidgets/rich-text-web/src/components/RichText.tsx:
```diff
@@ -0,0 +1,112 @@
+              width: "100%",
+              height: "100%"
+          };
+    const [ckeditorConfig, setCkeditorConfig] = useState<CKEditorHookProps<"change">>({
```
This `CKEditorHookProps<"change">` seems a bit strange, sounds like an on change handler or something 🤔

raffiar replied:
it's not controlled by me :) 
for more details check [CKEditor4 docs and react typings](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_react_v2.html#api-2)

raffiar commented on packages/pluggableWidgets/rich-text-web/rollup.config.js:
```diff
@@ -0,0 +1,51 @@
+function copyCKEditorDirToDist(outDir) {
+    return command([
+        async () => {
+            return copy("../../../node_modules/ckeditor4/", outDir, {
```
`resolve` returns path from HOME_Dir, but I believe `copy` is looking from the place where it started, so, I tried but it didn't give a result. No items have been copied to outDir :( 
Let me know if I missed something

diego-antonelli replied:
My bad, the correct one should be resolve. Same as https://github.com/mendix/widgets-resources/blob/3dea5d093148b8e7fc1b26ed364b8a42036e013d/packages/tools/pluggable-widgets-tools/configs/rollup-plugin-collect-dependencies.js#L7

raffiar committed:
feat(rich-text-web): fix pr comments

revert wrong master pull

raffiar commented on packages/pluggableWidgets/rich-text-web/src/components/__tests__/__snapshots__/RichText.spec.tsx.snap:
```diff
@@ -0,0 +1,12 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`RichText render DOM structure 1`] = `
```
This snapshot was removed

raffiar committed:
feat(rich-text-web): added mx glob variable

feat(rich-text-web): update package-lock

feat(rich-text-web): use resolve for path definition

feat(rich-text-web): use resolve for path definition

fix(rich-text-web): fix linting problems

feat(rich-text-web): fix broken unit tests

Keraito reviewed (changes requested):
Blockers are for advanced properties in Studio and the snapshot test for backwards compatibility going forward.

raffiar committed:
fix(rich-text-web): advanced mode visibility

Co-authored-by: Chak Shun Yu <chakshunyu@gmail.com>

raffiar committed:
feat(rich-text-web): added missing icons, styles

feat(rich-text-web): remove commented code

feat(rich-text-web): fix unit tests, added snapshot testing

Keraito reviewed (dismissed):
Latest commits look fine

raffiar committed:
fix(rich-text-web): fix broken event action when value was not available

raffiar committed:
feat(rich-text-web): added a spinner when value still not available

raffiar committed:
fix(rich-text-web): fix linting issues

raffiar committed:
feat(rich-text-web): added rich text icons
